src/widget/wspinnybase.cpp

```diff
@@ -0,0 +1,666 @@
+            // use xwax's bitmap to set alpha data only
+            // adjust alpha by 3/4 so it's not quite so distracting
+            // setpixel is slow, use scanlines instead
+            // m_qImage.setPixel(x, y, qRgba(r,g,b,(int)buf[x+m_iVinylScopeSize*y] * .75));
+            *line = qRgba(r, g, b, static_cast<int>(report.scope[x + m_iVinylScopeSize * y] * .75));
+            line++;
+        }
+    }
+#else
+    Q_UNUSED(report);
+#endif
+}
+
+void WSpinnyBase::setup(const QDomNode& node,
+        const SkinContext& context,
+        const ConfigKey& showCoverConfigKey) {
+    // Set images
+    QDomElement backPathElement = context.selectElement(node, "PathBackground");
+    m_pBgImage = WImageStore::getImage(context.getPixmapSource(backPathElement),
+            context.getScaleFactor());
+    Paintable::DrawMode bgmode = context.selectScaleMode(backPathElement,
+            Paintable::FIXED);
+    if (m_pBgImage && !m_pBgImage->isNull() && bgmode == Paintable::FIXED) {
+        setFixedSize(m_pBgImage->size());
+    } else {
+        setSizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::MinimumExpanding);
+    }
+    m_pMaskImage = WImageStore::getImage(
+            context.getPixmapSource(context.selectNode(node, "PathMask")),
+            context.getScaleFactor());
+    m_pFgImage = WImageStore::getImage(
+            context.getPixmapSource(context.selectNode(node, "PathForeground")),
+            context.getScaleFactor());
+    if (m_pFgImage && !m_pFgImage->isNull()) {
+        m_fgImageScaled = m_pFgImage->scaled(
+                size(), Qt::KeepAspectRatio, Qt::SmoothTransformation);
+    }
+    m_pGhostImage = WImageStore::getImage(
+            context.getPixmapSource(context.selectNode(node, "PathGhost")),
+            context.getScaleFactor());
+    if (m_pGhostImage && !m_pGhostImage->isNull()) {
+        m_ghostImageScaled = m_pGhostImage->scaled(
+                size(), Qt::KeepAspectRatio, Qt::SmoothTransformation);
+    }
+
+    // Dynamic skin option, set in WSpinnyBase's <ShowCoverControl> node.
+    if (showCoverConfigKey.isValid()) {
+        m_pShowCoverProxy = new ControlProxy(
+                showCoverConfigKey, this);
+        m_pShowCoverProxy->connectValueChanged(
+                this,
+                [this](double v) {
+                    m_bShowCover = v > 0.0;
+                });
+        m_bShowCover = m_pShowCoverProxy->get() > 0.0;
+    } else {
+        m_bShowCover = context.selectBool(node, "ShowCover", false);
+    }
+
+#ifdef __VINYLCONTROL__
+    // Find the vinyl input we should listen to reports about.
+    if (m_pVCManager) {
+        m_iVinylInput = m_pVCManager->vinylInputFromGroup(m_group);
+    }
+    m_iVinylScopeSize = MIXXX_VINYL_SCOPE_SIZE;
+    m_qImage = QImage(m_iVinylScopeSize, m_iVinylScopeSize, QImage::Format_ARGB32);
+    // fill with transparent black
+    m_qImage.fill(qRgba(0, 0, 0, 0));
+#endif
+
+    m_pPlayPos = new ControlProxy(
```

Comment:
here you can make use of `make_parented<ControlProxy>(` which asserts for a parent.

Reply:
Or even better use the light weight `PollingControlProxy` in the object memory which is already a special SmartPointer.

Reply:
This is not part of my code, this comes from the original WSpinny that has moved to WSpinnyBase

Reply:
Ok than it is fine.